apps/labs/posts/numpy-python-api-cleanup.md

```diff
@@ -0,0 +1,345 @@
+
+### Data type aliases analysis
+
+NEP 52 was also an opportunity to spend some time revisiting dtype aliases that are
+available in the main namespace `np.*` (1) and through `np.dtype(...)` (2).
+The debate on the final form is still ongoing (specifically `np.int_` and `np.uint`) but
+there are several groups of names that are available through (1) or (2):
+
+* words: `float`, `cdouble`, `uint`, `bool`, `object`, `long` ...
+* words+bits: `int16`, `uint8`, `float64`, ...
+* symbols: `p`, `L`, `i`, ...
+* symbols+bytes: `c8`, `i1`, `i2`, ...
+* abstract types: `numeric`, `inexact`, `integer`, ...
+
+The canonical names, that are available in the main namespace and should be used as
+a first choice, are "words+bits". These names are unambiguous about which type we're
+referring to, and we make an explicit declaration about precision, leaving no room for
+platform-specific behavior. "Words" that refer to C types should be used when
+interacting with C exctension modules directly.
+
+### Ruff plugin
+
+Ruff is a new Python linter that can outperform other well-known linting tools.
+During one of the discussions on Scientific Python Discord server, we came up with
+the idea of writing a dedicated Ruff rule to address NumPy 2.0 changes. The rule offers
+an automated way to fix a large part of changes - migrating to retained aliases or
+flagging lines that require manual intervention to be compatible with NumPy 2.0.
+
+This assignment gave me the opportunity to write a bit of Rust code, since it is
+the core language for Ruff. As of today, the PR is still open, but the rule is available
+in the local setup and it managed to correctly fix over a hundred of lines of code for
+the latest SciPy release source code.
+
+The command for running Ruff linter:
+
+```sh
+ruff check scipy/ --no-cache --fix
+```
+
+And here's a small passage from `git diff`:
+
+```diff
+-     	a = toarray(a, dtype=np.float_)
++     	a = toarray(a, dtype=np.float64)
+...
+-	elif (b == Inf and a == -Inf):
++	elif (b == np.inf and a == -np.inf):
+...
+-	math_dtypes = [np.int_, np.float_, np.complex_]
++	math_dtypes = [np.int_, np.float64, np.complex128]
+```
+
+Ruff rules open more robust ways to apply systematic changes to large codebases,
+compared to regular expression searches, because they offer AST analysis rather
+than a text-based search only.
+
+---
+
+## Wrapping up
+
+The development and maintenance efforts for NumPy package will continue, most notably
+in the Array API area, where full compatibility is still being implemented. The release
+of NumPy 2.0 is planned for end of the year, and only then will the official adoption
+of the new, major version begin.
+
+---
+
+## Acknowledgements
+
+I would like to thank my mentors [Ralf Gommers](https://github.com/rgommers) and
+[Nathan Goldbaum](https://github.com/ngoldbaum) for their advices and guidance during
+the whole internship, and [Melissa Weber Mendonça](https://github.com/melissawm) for
```

Comment:
Probably good to thank Sebastian as well? He did quite a bit of review.

Reply:
That's right - done